### PR TITLE
Export Simbody targets in SimbodyConfig.cmake files

### PR DESCRIPTION
**Pull request by elen4**
Currently, SimbodyConfig.cmake is an adaptation of FindSimbody.cmake, and libraries are found using the `find_library` command.
With this PR, libraries are *exported* so that downstream packages can import them (along with their properties) through SimbodyConfig.cmake.

Tested with OpenSim (master) compilation on Windows (VS2013, 64bit) and Ubuntu 14.04, but it's hard to tell if this can break other downstream packages' build :confused: 

**Note1**: There are also some pure-cleanup commits, hope they don't create too much confusion

**Note2**: I was trying also to get rid of the need for the deprecated `link_directories` command, but I cannot find a way to trick CMake into treating as targets the blas/lapack/pthread libraries that are provided as binaries on Windows :confounded:

### CODE REVIEW

**Comment by chrisdembia** [comment]
@sherm1 FYI Elena came by Clark yesterday and I talked with her for a bit about these changes. This change allows clients of Simbody to use commands like `target_link_libraries(mylib SimTKcommon)` instead of `target_link_libraries(mylib ${Simbody_LIBRARIES})`.

**Comment by chrisdembia** [comment]
@elen4 I'll do some of my own testing to make sure that Simbody still works well with client projects.

**Comment by chrisdembia** [comment]
@elen4 do you think the last line of [SampleCMakeLists.txt](https://github.com/simbody/simbody/blob/master/cmake/SampleCMakeLists.txt) should be updated? I am not sure what I think; we can leave it as is or we can update it; either way sounds fine.

File: CMakeLists.txt [hunk 1]
```diff
@@ -590,22 +590,33 @@ CONFIGURE_PACKAGE_CONFIG_FILE(
...
+        ${CMAKE_CURRENT_BINARY_DIR}/cmake/SimbodyConfigVersion.cmake
+        VERSION "${SIMBODY_VERSION}"
+        COMPATIBILITY SameMajorVersion)
+else()
+    include(CMakePackageConfigHelpers)
+     WRITE_BASIC_PACKAGE_VERSION_FILE(
+        ${CMAKE_CURRENT_BINARY_DIR}/cmake/SimbodyConfigVersion.cmake
+        VERSION "${SIMBODY_VERSION}"
+        COMPATIBILITY SameMajorVersion)
+endif()
```
Comment (chrisdembia) [inline CMakeLists.txt]: Older CMake versions are no longer supported. Can you leave only the second branch? Also, update the comment above to say "Requires CMake 2.8.8.".

**Comment by scpeters** [comment]
I'll test with gazebo this morning.

**Comment by scpeters** [comment]
Looks ok so far, still testing...

File: cmake/SimbodyConfig.cmake.in [hunk 2]
```diff
@@ -70,152 +63,13 @@ if (NOT "@SIMBODY_DOXYGEN_TAGFILE_NAME@" STREQUAL "")
...
-        "Simbody static libraries" FORCE)
-endif()
-
-# This CMake-supplied script provides standard error handling.
-include(FindPackageHandleStandardArgs OPTIONAL)
-find_package_handle_standard_args(Simbody DEFAULT_MSG 
-    Simbody_INCLUDE_DIR)
+# These are IMPORTED targets created by SimbodyTargets.cmake
+set(Simbody_LIBRARIES @SimTKCOMMON_SHARED_LIBRARY@ @SimTKMATH_SHARED_LIBRARY@ @SimTKSIMBODY_SHARED_LIBRARY@)
+set(Simbody_STATIC_LIBRARIES @SimTKCOMMON_STATIC_LIBRARY@ @SimTKMATH_STATIC_LIBRARY@ @SimTKSIMBODY_STATIC_LIBRARY@)
```
Comment (sherm1) [inline cmake/SimbodyConfig.cmake.in]: Wow -- this is really nice! I love seeing all that code deleted.

**Comment by elen4** [comment]
@chrisdembia I updated the VersionConfig-related code (forgot to do that yesterday :disappointed: ). I edited the related commit and push-forced to keep the history clean, hope it didn't cause problems with your testing

I would keep SampleCMakeLists.txt as it is, IMHO it is meant for beginners who wouldn't know which libraries they need to link against anyway, so it's safer if they just link them all. Or maybe you can add a comment with the alternative `target_link_libraries(myexe SimTKcommon)`  (or similar) command, so that more advanced users know that they can use it :smile:

**Comment by chrisdembia** [comment]
> IMHO it is meant for beginners who wouldn't know which libraries they need to link against anyway, so it's safer if they just link them all.

K I agree.

File: cmake/SimbodyConfig.cmake.in (hunk 2 above)
Comment (chrisdembia) [inline cmake/SimbodyConfig.cmake.in]: With this PR, Simbody_STATIC_LIBRARIES is defined even if the user did not build static libraries. Likewise, Simbody_LIBRARIES is defined even if shared libraries are not built. Is this desired? I think this behavior is different from before.

**Comment by sherm1** [comment]
I tested this on Windows and it is no longer able to find the simbody-visualizer when installed in an arbitrary location. That's because it used to build with the full installation path as one of the entries in the visualizer search path. The switch to relative paths means that entry is now just "/bin/". Here is the error message:
```
C:\Program Files\Simbody-elen4\examples\bin>examplecablepath
EXCEPTION: SimTK Exception thrown at VisualizerProtocol.cpp:157:
  Error detected by Simbody method VisualizerProtocol::ctor(): Unable to spawn executable 'simbody-visualizer' from directories:
  c:\Program Files\Simbody-elen4\examples\bin\
  c:\Program Files\Simbody\bin\
  bin/
  C:\Program Files\Simbody\bin\
  C:\Program Files\SimTK\bin\
Final system error was errno=2 (No such file or directory).
  (Required condition 'status == 0' was not met.)
```

Previous (long) discussions about this can be found in #339 and especially #64 with Thomas Moulard from Debian.

File: CMakeLists.txt [hunk 3]
```diff
@@ -523,13 +523,13 @@ LINK_DIRECTORIES(${BUILD_BINARY_DIR})
...
   # Install visualizer to bin, since it needs to be co-located with dll's
-  SET(SIMBODY_VISUALIZER_INSTALL_DIR ${CMAKE_INSTALL_FULL_BINDIR})
+  SET(SIMBODY_VISUALIZER_INSTALL_DIR ${CMAKE_INSTALL_BINDIR})
   # Install include files into base include folder since it's a sandbox
   SET(SIMBODY_INCLUDE_INSTALL_DIR ${CMAKE_INSTALL_INCLUDEDIR})
 ELSE()
   # Visualizer is not intended to be a user executable. Proper place is
   # inside the lib directory
-  SET(SIMBODY_VISUALIZER_INSTALL_DIR ${CMAKE_INSTALL_FULL_LIBEXECDIR}/simbody)
+  SET(SIMBODY_VISUALIZER_INSTALL_DIR ${CMAKE_INSTALL_LIBEXECDIR}/simbody)
```
Comment (sherm1) [inline CMakeLists.txt]: I believe this change and the similar one just above are why the visualizer search doesn't work now.

Comment (chrisdembia) [inline CMakeLists.txt]: Ah good point.

**Comment by chrisdembia** [comment]
> I tested this on Windows and it is no longer able to find the simbody-visualizer when installed in an arbitrary location.

Linux also

File: CMakeLists.txt (hunk 3 above)
Comment (elen4) [inline CMakeLists.txt]: If I revert the changes to these two lines, does it work? I need more time to figure out how the visualizer-related stuff works :sweat: I'm not sure if the package will be relocatable in the end...

Comment (chrisdembia) [inline CMakeLists.txt]: > I'm not sure if the package will be relocatable in the end...

Don't worry about this for your PR. The package is still relocatable but would then require that the simbody-visualizer is on the path...I think.

Comment (chrisdembia) [inline CMakeLists.txt]: Let me clarify: the package is still relocatable but would then require that SIMBODY_HOME is set correctly.

File: cmake/SimbodyConfig.cmake.in (hunk 2 above)
Comment (elen4) [inline cmake/SimbodyConfig.cmake.in]: I had wrongly assumed that variables like `SimTKCOMMON_STATIC_LIBRARY` would only be defined if `BUILD_STATIC_LIBRARY` was ON, but this is not the case. I can (change all the CMake files or ) wrap the `set(Simbody_STATIC_LIBRARIES ...` line in a `if(@BUILD_STATIC_LIBRARIES@)` block. I don't think you can disable compilation of shared libraries.

Anyway, this is a really tricky point: how is client code supposed to deal with the un/availability of static libraries? Does it just check if `Simbody_STATIC_LIBRARIES` is defined? I will try to see if there is a way for client code to specify which libraries they want in the `find_package` call

Comment (sherm1) [inline cmake/SimbodyConfig.cmake.in]: > how is client code supposed to deal with the un/availability of static libraries? 

I don't think this is a big issue in practice. Everyone uses shared libraries except in special cases where they must have static libraries. In that case they will just have to make sure the static libraries are available.

File: CMakeLists.txt (hunk 3 above)
Comment (sherm1) [inline CMakeLists.txt]: > the package is still relocatable but would then require that SIMBODY_HOME is set correctly

That's right -- the intention was that it would always be able to find the visualizer with no environment variable set as long as the package was installed in the directory specified in CMake at the time it was built. If the installation were moved somewhere else, then it would still work but would require `SIMBODY_HOME` to be set.

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -67,7 +67,7 @@ IF(WIN32)
   SET(EXAMPLES_INSTALL_BIN examples/bin/)
   SET(EXAMPLES_INSTALL_SRC examples/src/)
 ELSE()
-  SET(EXAMPLES_INSTALL_BIN ${CMAKE_INSTALL_FULL_LIBDIR}/simbody/examples/)
+  SET(EXAMPLES_INSTALL_BIN ${CMAKE_INSTALL_LIBDIR}/simbody/examples/)
```
Comment (chrisdembia) [inline examples/CMakeLists.txt]: I now realize why this was FULL_LIBDIR: this variable is used to create a symlink, and the symlink must be given the full path. See [here](https://github.com/elen4/simbody/blob/simbodyConfig/examples/CMakeLists.txt#L157); perhaps prepend `CMAKE_INSTALL_PREFIX` here? The symlink will be broken if the user moves the installation but I think that's fine.

**Comment by chrisdembia** [comment]
I did some testing on Ubuntu 14.04:

- I tested installing RelWithDebInfo and Debug on top of each other; if my client project's build type is Debug, then Simbody's debug libraries are used. :heavy_check_mark: 
- I checked the values of all the variables that SimbodyConfig.cmake sets; they are all correct. :heavy_check_mark:

**Comment by sherm1** [comment]
I think we might need a checklist of tests for this PR before merging it. Some that come to mind:
- need to test all of these on Windows, Linux, OSX make, OSX homebrew
- should be able to find visualizer when installed in `CMAKE_INSTALL_PREFIX`, with no environment variable set
- examples should run directly out of build directory when building (using just-built libraries), and directly out of install directory 
- should be able to build examples elsewhere and run them (that means they have to be able to locate auxiliary files that are still in the install examples/src directory). Symbolic links get created for that on Linux & OSX
- everything should still work if the installation is moved elsewhere provided `SIMBODY_HOME` is set

I'm not sure if that is exhaustive -- are there other things that might break?

**Comment by chrisdembia** [comment]
Some more:

- Gazebo should require no changes to the build process for using Simbody.
- OpenSim should require no changes to the build process for using Simbody.
- The variables that SimbodyConfig.cmake defines are all correct (build doxygen as well).
- Logical behavior when client project wants to use Simbody's static libraries but they were not built.
- Successful build of client project when client project wants to use Simbody's static libraries and they WERE built.

Comments on the above:



> everything should still work if the installation is moved elsewhere provided SIMBODY_HOME is set

This is currently not true, since VisualizerProtocol.cpp looks in SIMBODY_HOME/bin even on Linux, but the serach path should be SIMBODY_HOME/libexec/simbody; I'm working on a separate PR for this. Also, this checklist item I think is unrelated to the current PR.



> Symbolic links get created for that on Linux & OSX

I'm not sure which symbolic links you're referring to. Auxiliary files are found using logic in SimbodyExampleHelper.h, not with symlinks. There's only one symlink I think for locating the example binaries from /share.

**Comment by sherm1** [comment]
> If I revert the changes to these two lines, does it work? 

I tried changing the two `SIMBODY_VISUALIZER_INSTALL_DIR` lines back to using `_FULL` and then retested on Windows. In that case it was able to find the visualizer, so I think that fixes it. But will that break whatever else it was you were trying to accomplish?

**Comment by chrisdembia** [comment]
> But will that break whatever else it was you were trying to accomplish?

Not really. Elena tried to remove the full paths wherever they appeared. In some cases, the full path really made the installation non-relocatable (under Elena's *new* SimbodyConfig.cmake); the two changes you just made are fine.

**Comment by sherm1** [comment]
> the two changes you just made are fine.

I only made them locally on my machine, not in Elena's fork.

**Comment by elen4** [comment]
I changed the relevant commit so that `SIMBODY_VISUALIZER_INSTALL_DIR` is left untouched. I don't really understand the rationale of the symlink in examples/CMakeLists.txt, I will try to look into it further. In the meantime I can revert that change as well, it was not strictly related to this PR as it does not involve targets that get exported; as Chris said, I was just removing all FULL variables as I didn't see a reason for them and I was trying to keep the CMakeLists consistent.

**Comment by sherm1** [comment]
Hi, Elena. On Linux there is a rule that machine-dependent files (like example executable binaries) can't be in `share` directories, which are presumed to be shared among dissimilar computers that would need different binaries. The examples are considered documentation so they are included in `/usr/local/share/doc/simbody/examples`. That makes sense for the source files, but isn't allowed for the binaries. But it also makes sense to have the binaries and source together, so we use a symbolic link for the `bin` directory that references `/usr/local/lib/simbody/examples` which is a machine-specific location.

**Comment by elen4** [comment]
Hi Sherm, thanks for the explanation! I now understand your reasons, yet I find the current implementation quite glitchy. Not only is the final package not relocatable, but I found that if I changed the installation directory after the first CMake pass, the symbolic link would still point to the first (default) installation directory. I committed [here](https://github.com/elen4/simbody/commit/7fa84f2a6703642f232e959fdfda345721ba0ba4) an alternative solution, in which the symbolic link is created at installation time, and it contains a relative path so that the resulting installation is relocatable. What do you think? Should I remove the change to the examples CMakeLists from commit 8ff1622d0048 and open a different pull request?

**Comment by sherm1** [comment]
I'm not sure that it makes sense to think of the current Linux installation as relocatable, because it is spread around so many different directories. I don't think there is any reasonable way for anyone to relocate it. I think this is even more complicated on OSX-Homebrew where the actual installation is in `Cellar` and then symlinks are made for everything else. This would require a scary amount of careful testing on every platform; it took a long time to get it working as it is now.

What I think would be more useful would be an alternate self-contained installation for Linux and OSX where everything would go under `/usr/local/simbody` (or wherever), similar to how installation works on Windows. Then making it relocatable would make sense, but there wouldn't need to be any symbolic links in that case. I would envision this style of installation as a CMake option chosen at build time, or maybe an addition target like `make install` and `make install-self-contained` (or whatever the right terminology would be). I'm not sure how many people would use this style of installation but I personally find it annoying to have the installation spread around so much. Do you have an opinion, Chris?

As a practical matter, I think the build/install system is now complicated enough that we can't make substantive changes to it without testing it carefully on all the supported platforms. Currently that means: Windows (relocatable), Linux (not relocatable; must meet Debian requirements), OSX (Linux-style), OSX (Homebrew). This testing is the most difficult part of this PR, and I don't know how it can be automated. It may make more sense to separate this into a "cleanup" PR that doesn't change the functionality, and then treat functionality changes as smaller, separate PRs that will require more testing.

**Comment by sherm1** [comment]
> if I changed the installation directory after the first CMake pass, the symbolic link would still point to the first (default) installation directory

Good find! That sounds like a bug that should be fixed. Doing it in its own PR would make it easier to test since we would know the limited circumstances in which the change has an effect.

**Comment by chrisdembia** [comment]
> Do you have an opinion, Chris?

I don't think relocatability of the FHS-compliant installation (on Linux) is out of reach; I'm not sure what else would need to change other than the symlink. However, I agree that making the FHS-compliant installation relocatable is not so useful, and so the effort to test that the FHS-compliant installation is relocatable is not worthwhile.

I think that there should be a CMake option to use a Windows-style (relocatable) or UNIX-style installation (not relocatable). So on Linux, if you want to be able to relocate the installation, use the Windows style installation.

I agree that this PR should focus on the new SimbodyConfig.cmake content, and so should roll back to change to the symlink path. We can worry about relocatability and a potentially new CMake variable later. Basically, I agree with Sherm.

**Comment by elen4** [comment]
I removed changes to the examples CMakeLists from this PR, and I modified the way that `Simbody_STATIC_LIBRARIES` is set. I tested it on Windows with the following client CMakeLists file:

``` cmake
cmake_minimum_required(VERSION 2.8.8)
find_package(Simbody REQUIRED)
message("Simbody_STATIC_LIBRARIES: ${Simbody_STATIC_LIBRARIES}")
if (Simbody_STATIC_LIBRARIES)
  message("Static libraries were built")
else()
  message("Static libraries were NOT built")
endif()
```

and it seems to work, not sure if this is how current CMake code handles the availability of static libraries. @chrisdembia do you think this will work?

**Comment by chrisdembia** [comment]
> @chrisdembia do you think this will work?

Yes! Thank you.

**Comment by chrisdembia** [comment]
I tested this PR on OSX (make) and Windows. 

- Examples run directly out of the build directory when building.
- To run examples from the install directory, DYLD_LIBRARY_PATH must be set on OSX, and PATH must be set on Windows. After setting these paths, the examples work (including finding geometry and visualizer).
- Building a client project (including OpenSim) goes smoothly. 
- On both platforms, to run my client project's executable (from its own build directory) I must add the location of the Simbody libraries to either DYLD_LIBRARY_PATH or PATH (not so for opensim, since opensim copies the libraries to the binary dir). I think this was necessary *before* this PR, though.
- The `BUILD_USING_NAMESPACE` functionality still works (only tested on OSX).
- Building and running the examples elsewhere works (after setting DYLD_LIBRARY_PATH or PATH), but uses the geometry from "elsewhere", not from the simbody install. I think this is preferable though. Also, I doubt this behavior changed with this PR. Visualizer is located successfully.
- If a client project requests Simbody static libraries and they haven't been built, you get an error (which is the desired behavior). If they *have* been built, then they are used. That's on OSX. With windows, trying to use Simbody static libraries to build the examples ("elsewhere") gives linker errors (with or without this PR, so this PR does not introduce this issue).

I did not test relocatability on any platform. As we discussed above, it doesn't seem like we think this is a reasonable objective, at least not for this PR.

I also did not test gazebo, but I don't expect any differences there.

I think we can merge this. @elen4 can you resolve the merge conflicts?

**Comment by elen4** [comment]
@chrisdembia do you prefer that I rebase all the commits on master and force-push (keeps history cleaner, and in this case is not hard to do, just some whitespace deletion to fixup), or that I merge master into this branch?

**Comment by sherm1** [comment]
Elena, we are relaxed about the contents of history. Merging master in is fine. Thanks!

**Comment by chrisdembia** [comment]
Yeah I think the dirty history is better than the risks from rewritten history caused by force-pushing (though sometimes I do it :) )

**Comment by elen4** [comment]
I merged master in, let me know if there is anything else!

**Comment by sherm1** [comment]
> I merged master in, let me know if there is anything else!

Are there any documentation changes needed to match this? Please take a look at [README.md](https://github.com/simbody/simbody/blob/master/README.md) to see whether it is out of date now. I think it should have a short section explaining how you would advise people to make use of the new installed cmake stuff to build their own Simbody-using programs. Currently there is a "Using Simbody" section but it is almost empty.

**Comment by sherm1** [comment]
I pulled this branch onto my Windows machine to try it out but got a build failure, which is strange since the build on AppVeyor worked. I'm using CMake 3.0.2 and Visual Studio 2013. The errors were all in VisualizerProtocol.cpp. Any thoughts?
```
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(261): error C2065: 'SIMBODY_VISUALIZER_REL_INSTALL_DIR' : undeclared identifier
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(266): error C2065: 'SIMBODY_VISUALIZER_REL_INSTALL_DIR' : undeclared identifier
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(270): error C2143: syntax error : missing ')' before ';'
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(270): error C2065: 'SIMBODY_VISUALIZER_REL_INSTALL_DIR' : undeclared identifier
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(270): error C2059: syntax error : ')'
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(279): error C2065: 'SIMBODY_VISUALIZER_REL_INSTALL_DIR' : undeclared identifier
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(279): error C2660: 'SimTK::Pathname::addDirectoryOffset' : function does not take 1 arguments
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(282): error C2065: 'SIMBODY_VISUALIZER_REL_INSTALL_DIR' : undeclared identifier
32>C:\Users\Sherm\Documents\GitHub\simbody\simbody-master\Simbody\Visualizer\src\VisualizerProtocol.cpp(282): error C2660: 'SimTK::Pathname::addDirectoryOffset' : function does not take 1 arguments
```

**Comment by sherm1** [comment]
Oh, I see the continuous integration build on AppVeyor specifies `BUILD_VISUALIZER=OFF` so doesn't exercise this code.

**Comment by sherm1** [comment]
Never mind my last comment -- I'm still getting the failure with `BUILD_VISUALIZER=OFF` since VisualizerProtocol.cpp is built in the Simbody library regardless.

**Comment by sherm1** [comment]
Problem turned out to be a typo in the installation directory name I had chosen in CMake: `C:/Temp[/SimbodyElen4`. Somehow that extra left bracket caused the compilation failure. Back to testing ...

**Comment by elen4** [comment]
@sherm1 regarding the documentation (i.e., the SampleCMakeLists), I prepared a gist with an updated version with tips for smart users: https://gist.github.com/elen4/6ceddf131e2badac61f7
Then, randomly scrolling this page, I saw that @chrisdembia brought it up on April 3rd and I commented:
> I would keep SampleCMakeLists.txt as it is, IMHO it is meant for beginners who wouldn't know which libraries they need to link against anyway, so it's safer if they just link them all. Or maybe you can add a comment with the alternative target_link_libraries(myexe SimTKcommon) (or similar) command, so that more advanced users know that they can use it :smile: 

Let me know which version you prefer (current vs gist). Also, thanks for testing, and :relieved: for the typo! Not sure if this corner case should be addressed somehow, but I don't think it's related to this PR.

**Comment by sherm1** [comment]
Thanks, Elena. I don't want to encourage people to link with individual libraries because there is no particular utility to having those separated out, and no guarantee that that the code will continue to be split up that way. I have been thinking about combining them into a single libsimbody but haven't got around to it yet. So I think the current SampleCMakeLists.txt is OK. I'll merge this now.
